### PR TITLE
JSON schema updates

### LINKED ISSUE #7823: JSON Schema for composer.lock file
Hi,

there appears to be no JSON schema for composer.lock structure.
This is a bit problematic for tools building on top of the lockfile, be it for reading, manipulation, maintaining bi-directional compatibility or different use cases.

Would it be possible to add one?

Thanks.

### PR DESCRIPTION
**Pull request by Seldaek**
- Added composer-lock-schema
- Updated composer-repository-schema with new properties
- Added lock file schema validation in `diagnose`

Fixes #7823

### CODE REVIEW

File: res/composer-lock-schema.json [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+                "type": { "type": "string" },
+                "version": { "type": "string" },
+                "version_normalized": {
+                    "type": "string",
+                    "description": "Normalized version, optional but can save computational time on client side."
+                },
+                "autoload": { "type": "object" },
+                "require": { "type": "object" },
+                "replace": { "type": "object" },
+                "conflict": { "type": "object" },
```
Comment (stof) [inline res/composer-lock-schema.json:121]: should those define the `additionalProperties` as being strings ?

File: res/composer-lock-schema.json [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+        "package-base": {
+            "properties": {
+                "name": { "type": "string" },
+                "type": { "type": "string" },
+                "version": { "type": "string" },
+                "version_normalized": {
+                    "type": "string",
+                    "description": "Normalized version, optional but can save computational time on client side."
+                },
+                "autoload": { "type": "object" },
```
Comment (stof) [inline res/composer-lock-schema.json:118]: should this define the more precise type ?

File: res/composer-lock-schema.json (hunk 1 above)
Comment (Seldaek) [reply res/composer-lock-schema.json:121]: Yeah if we want to be more specific we could perhaps refer to the composer-schema.json for the entire packages contents. That'd be better but I just took this from the repo schema for now which is definitely not as complete.

On the other hand refs across files makes it a bit harder to use standalone so I am thinking it's probably good enough as is.. Ultimately the main point is to validate the lock file specific parts. Each package is then just a composer.json.

Comment (naderman) [reply res/composer-lock-schema.json:121]: Think if we're adding this schema to make it easier to work with this file we should either make it accurate and list everything out here, or use the reference if you can't be bothered to copy stuff / maintain it in two places, but just waving it off with type object here seems like only doing half the task? ;-P

Comment (boesing) [reply res/composer-lock-schema.json:121]: Maybe create a version which refers to the file and the dereference with a tool to the expected location in a build? That would be best of both worlds.

Comment (Seldaek) [reply res/composer-lock-schema.json]: Ok I've added a ref.. It seems to work well enough so let's keep it simple for now :)

**Review by stof — APPROVED** [review]

**Pull request merged**